### PR TITLE
Apply Infinispan CR labels changes to the Service,Route,Ingress and Pod resources

### PR DESCRIPTION
**Pull request by dmvolod**

### CODE REVIEW

**Comment by dmvolod** [comment]
@ryanemerson please pay attention that 12.0.1.Final image didn't solve floating `404 Not Found` error in Backup/Restore test

**Comment by ryanemerson** [comment]
> @ryanemerson please pay attention that 12.0.1.Final image didn't solve floating `404 Not Found` error in Backup/Restore test

The combination of `12.0.1.Final` and  changes in https://github.com/infinispan/infinispan-operator/pull/770 should resolve the 404 issues.

File: pkg/apis/infinispan/v1/types_util.go
Comment (ryanemerson) [inline pkg/apis/infinispan/v1/types_util.go:435]: Is the "\n" character necessary?

File: pkg/apis/infinispan/v1/types_util.go
Comment (ryanemerson) [inline pkg/apis/infinispan/v1/types_util.go:431]: Is the "\n" character necessary?

File: pkg/controller/infinispan/infinispan_controller.go
Comment (ryanemerson) [inline pkg/controller/infinispan/infinispan_controller.go:659]: Why are we updating pod's individually instead of just updating the StatefulSet spec?

File: pkg/controller/infinispan/infinispan_controller.go
Comment (ryanemerson) [inline pkg/controller/infinispan/infinispan_controller.go:657]: @rigazilla Is there any reason we can't add these calls inside `PodLabels` so that it doesn't have to be repeated in multiple places?

**Comment by dmvolod** [comment]
> The combination of `12.0.1.Final` and changes in #770 should resolve the 404 issues.

Good news, will wait until #770 be merged

File: pkg/controller/infinispan/infinispan_controller.go
Comment (ryanemerson) [inline pkg/controller/infinispan/infinispan_controller.go:687]: @rigazilla  Why do we need `lsPod` to be distinct from `labelsForPod`? Won't both result in a match as all of the labels will be on the pods?

File: pkg/controller/infinispan/resources/service/service_controller.go
Comment (dmvolod) [reply pkg/controller/infinispan/resources/service/service_controller.go:151]: Seems to be none as multiple foreign controllers could update Service and `CreateOrPatch` could conflict on `Patch` as this is non-blocking operation

File: pkg/controller/infinispan/resources/service/service_controller.go
Comment (dmvolod) [reply pkg/controller/infinispan/resources/service/service_controller.go:144]: Yes, sure. Missed that Route and Ingress are also labeled

File: pkg/apis/infinispan/v1/types_util.go
Comment (dmvolod) [reply pkg/apis/infinispan/v1/types_util.go:435]: I don't know as this code was originally implemented by @rigazilla . I just add a fix for using `fmt.Sprintf` for all vars

File: pkg/apis/infinispan/v1/types_util.go
Comment (dmvolod) [reply pkg/apis/infinispan/v1/types_util.go:431]: I don't know as this code was originally implemented by @rigazilla . I just add a fix for using `fmt.Sprintf` for all vars

File: pkg/controller/infinispan/infinispan_controller.go
Comment (dmvolod) [reply pkg/controller/infinispan/infinispan_controller.go:659]: Updating `Spec.PodTemplateSpec` will cause a rolling update of the whole cluster just for Pods updating that's necessary in our case

Comment (ryanemerson) [reply pkg/controller/infinispan/infinispan_controller.go:659]: As the update is only to the metadata labels, does this not cause the individual pods to be restarted?

Comment (dmvolod) [reply pkg/controller/infinispan/infinispan_controller.go:659]: Which update? Any StatefulSet `PodTemplateSpec` update including annotations or labels will cause rolling update of whole cluster. For individual Pods update only annotations or labels will be updated.

Comment (ryanemerson) [reply pkg/controller/infinispan/infinispan_controller.go:659]: Sorry I meant "As the update is only to the metadata labels of the *individual pods*, does this not cause the individual pods to be restarted?". But I guess not :slightly_smiling_face:

File: pkg/apis/infinispan/v1/types_util.go
Comment (rigazilla) [reply pkg/apis/infinispan/v1/types_util.go:438]: yeah it's not needed

File: pkg/controller/infinispan/infinispan_controller.go
Comment (rigazilla) [reply pkg/controller/infinispan/infinispan_controller.go:673]: I wanted to leave the possibility to update only CR or operator labels, I'm not sure if that case will ever happens anyway

File: pkg/apis/infinispan/v1/types_util.go
Comment (dmvolod) [reply pkg/apis/infinispan/v1/types_util.go:438]: Thanks, @rigazilla will remove trailing `\n`

File: pkg/controller/infinispan/resources/service/service_controller.go
Comment (ryanemerson) [inline pkg/controller/infinispan/resources/service/service_controller.go:218]: AFAICT all three reconcile functions are identical, except the object type being updated. Could we use `ToUnstructured` to create a generic method to handle the following part of the code:

```
		result, err := controllerutil.CreateOrUpdate(context.TODO(), s.client, findIngress, func() error {
		if findIngress.CreationTimestamp.IsZero() {
			controllerutil.SetControllerReference(s.infinispan, findIngress, s.scheme)
			findIngress.Spec = ingress.Spec
			findIngress.Annotations = ingress.Annotations
		}
		findIngress.Labels = ingress.Labels
		return nil
	})
	if err != nil {
		s.log.Error(err, "failed to create or update Ingress", ingress)
		return err
	}
	if result != controllerutil.OperationResultNone {
		s.log.Info(fmt.Sprintf("%s Ingress %s", strings.Title(string(result)), findIngress.GetName()))
	}
```

Comment (dmvolod) [reply pkg/controller/infinispan/resources/service/service_controller.go:218]: This could be an option, but I we would like to change some `Spec` structures not sure that will be a possible. For `ObjectMeta` it should be possible

File: pkg/controller/infinispan/infinispan_controller.go
Comment (rigazilla) [reply pkg/controller/infinispan/infinispan_controller.go:703]: The match will probably work, but I would not mix up internal labels `lsPod` with user labels `labelsForPod` which come for the user application domain and are not intended to be used as pods selector.

File: pkg/controller/infinispan/resources/service/service_controller.go
Comment (ryanemerson) [reply pkg/controller/infinispan/resources/service/service_controller.go:218]: I haven't tried it, but I think you could do something like:

```go
		found := unstructured.ToUnstructured(findIngress)
		provided := unstructured.ToUnstructured(ingress)
		unstructured.SetNestedField(found, provided["Spec"], "Spec")
		unstructured.SetNestedField(found, provided["Annotations"], "Annotations")
```
But maybe the saved lines aren't worth it :slightly_smiling_face:

Comment (dmvolod) [reply pkg/controller/infinispan/resources/service/service_controller.go:218]: Thanks for idea, @ryanemerson 
Bit polished and used for common function. But we should avoid to use `Unstructured` a lot.

Comment (dmvolod) [reply pkg/controller/infinispan/resources/service/service_controller.go:177]: Fixed, but Baclup/Restore test failed with another error

File: pkg/controller/infinispan/infinispan_controller.go
Comment (rigazilla) [inline pkg/controller/infinispan/infinispan_controller.go:690]: Does this wipe out all the labels from the pod and then restore only our stuff? If so, it's too invasive imo

**Review by rigazilla — CHANGES_REQUESTED** [review]

File: pkg/controller/infinispan/infinispan_controller.go
Comment (dmvolod) [reply pkg/controller/infinispan/infinispan_controller.go:690]: No, the system labels controlled by `consts.StatefulSetPodLabels` and never been removed

Comment (rigazilla) [reply pkg/controller/infinispan/infinispan_controller.go:690]: I'm referring to labels possibly added by the user for his own business, maybe I'm misreading the code but looks like they are not preserved.
also the `cordinator` label for cross site is removed if present.
I would apply a patch policy instead of rebuilding the labels map from scratch

Comment (dmvolod) [reply pkg/controller/infinispan/infinispan_controller.go:690]: Good catch related coordinator label for Pods, thanks @rigazilla . Will merge it logic with current update. Not sure that customers will need own labels on Pods controlled by statefulset, they will be removed every rolling upgrade. If customers will need own labels, they will need to set them via Infinispan CR.

Comment (dmvolod) [reply pkg/controller/infinispan/infinispan_controller.go:690]: > I would apply a patch policy instead of rebuilding the labels map from scratch

Also is unable to implement two-way merge without creating labels map from scratch, but it seems to be similar logic

Comment (dmvolod) [reply pkg/controller/infinispan/infinispan_controller.go]: @rigazilla fixed issue with coordinator Pod label

Comment (rigazilla) [reply pkg/controller/infinispan/infinispan_controller.go]: I still prefer my idea of not actively delete unknown labels, but that's just a detail, overall LGTM.

**Review by rigazilla — APPROVED** [review]

File: pkg/controller/infinispan/infinispan_controller.go
Comment (dmvolod) [reply pkg/controller/infinispan/infinispan_controller.go]: @rigazilla yes, sure, but I can't imagine how to remove user defined label which user was removed.

Comment (rigazilla) [reply pkg/controller/infinispan/infinispan_controller.go]: oh sorry @dmvolod I didn't though to that problem. I got your point now, your are right.

Actually I can something like: if the label name is listed in the annotations but not in the infinispan CR labels, then delete the label on the pod/svc.
this complicates things a bit and I'm not sure if it has side effects.

Comment (dmvolod) [reply pkg/controller/infinispan/infinispan_controller.go]: Yes, that's possible solution, but can't handle edit annotations.

Comment (rigazilla) [reply pkg/controller/infinispan/infinispan_controller.go]: yeah fine. I think we can go with this solution, let's keep overcomplicated solutions if users will complain about disappearing labels.

**Comment by ryanemerson** [comment]
Thanks @dmvolod and sorry for the delay.

**Pull request merged**
